### PR TITLE
User Tags in Instance, Inst Temp, Inst Vol Attmnt

### PR DESCRIPTION
**Pull request by SunithaGudisagarIBM**
### Community Note

* Please vote on this pull request by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original pull request comment to help the community and maintainers prioritize this request
* Please do not leave "+1" or other comments that do not add relevant new information or questions, they generate extra noise for pull request followers and do not help prioritize the request

Relates OR Closes #0000

Output from acceptance testing:

```
$ make testacc TEST=./ibm/service/vpc TESTARGS='-run=TestAccIBMISInstanceTemplate_WithVolumeAttachmentUserTag'
$ make testacc TEST=./ibm/service/vpc TESTARGS='-run=TestAccIBMISInstance_bootVolumeUserTags'
$ make testacc TEST=./ibm/service/vpc TESTARGS='-run=TestAccIBMISInstanceVolumeAttachment_userTag'
```

```
resource "ibm_is_instance" "testacc_instance" {
  name    = "sunitha-instance-new"
  image   = "r134-c31deded-722d-4950-9d05-862fecb831b1"
  profile = "cx2-2x4"
  boot_volume {
    size = 220
    tags = ["tag-0", "tag-1"]
  }
  primary_network_interface {
    subnet = "0716-4b83a399-a222-4736-a0e0-44cd802a7e64"
  }
  user_data = "a"
  vpc       = "r134-060fbfe6-4d0f-47b0-b9c5-94da8e719e22"
  zone      = "us-south-1"
  keys      = ["r134-8ddfdc65-73ac-48e1-864b-870f29bd35ca"]
}

resource "ibm_is_instance_volume_attachment" "testacc_att" {
  instance = ibm_is_instance.testacc_instance.id

  name     = "volume-attachment"
  profile  = "custom"
  capacity = 20
  iops     = 600
  tags     = ["tag-sunitha","test"]

  delete_volume_on_instance_delete = true
  volume_name                      = "sunitha-volume-inst-vol-attmnt"
}
```

```
sunitha@dhcp-9-239-8-182 vpc-ibm-lb-config % terraform apply -auto-approve
ibm_is_instance.testacc_instance: Refreshing state... [id=0716_c8a9c7f3-bbec-4994-87f1-21096566acd6]
ibm_is_instance_volume_attachment.testacc_att: Refreshing state... [id=0716_c8a9c7f3-bbec-4994-87f1-21096566acd6/0716-9bd05fef-6920-48f7-8c50-a9cea3d47ade]

Terraform used the selected providers to generate the following execution plan. Resource actions are indicated with the following symbols:
  ~ update in-place

Terraform will perform the following actions:

  # ibm_is_instance_volume_attachment.testacc_att will be updated in-place
  ~ resource "ibm_is_instance_volume_attachment" "testacc_att" {
        id                                 = "0716_c8a9c7f3-bbec-4994-87f1-21096566acd6/0716-9bd05fef-6920-48f7-8c50-a9cea3d47ade"
        name                               = "volume-attachment"
      ~ tags                               = [
          + "test",
            # (1 unchanged element hidden)
        ]
        # (15 unchanged attributes hidden)
    }

Plan: 0 to add, 1 to change, 0 to destroy.
ibm_is_instance_volume_attachment.testacc_att: Modifying... [id=0716_c8a9c7f3-bbec-4994-87f1-21096566acd6/0716-9bd05fef-6920-48f7-8c50-a9cea3d47ade]
ibm_is_instance_volume_attachment.testacc_att: Still modifying... [id=0716_c8a9c7f3-bbec-4994-87f1-21096566ac...6-9bd05fef-6920-48f7-8c50-a9cea3d47ade, 10s elapsed]
ibm_is_instance_volume_attachment.testacc_att: Still modifying... [id=0716_c8a9c7f3-bbec-4994-87f1-21096566ac...6-9bd05fef-6920-48f7-8c50-a9cea3d47ade, 20s elapsed]
ibm_is_instance_volume_attachment.testacc_att: Modifications complete after 27s [id=0716_c8a9c7f3-bbec-4994-87f1-21096566acd6/0716-9bd05fef-6920-48f7-8c50-a9cea3d47ade]

Apply complete! Resources: 0 added, 1 changed, 0 destroyed.
```

screen1
screen2
screen3

data "ibm_is_instance_template" "instance_template_data" {
name = "cli-tpl-1"
}
data "ibm_is_instance_templates" "instance_templates_data" {
}

screen5

screen4

### CODE REVIEW

**Comment by SunithaGudisagarIBM** [comment]
@hkantare we have reviewed internally this feature please find the PR https://github.com/ibm-vpc/terraform-provider-ibm/pull/162 here.. Can you please review and merge this PR...

File: ibm/service/vpc/resource_ibm_is_instance_template.go [hunk 1]
```diff
@@ -239,6 +242,14 @@ func ResourceIBMISInstanceTemplate() *schema.Resource {
 										ForceNew:    true,
 										Description: "The CRN of the [Key Protect Root Key](https://cloud.ibm.com/docs/key-protect?topic=key-protect-getting-started-tutorial) or [Hyper Protect Crypto Service Root Key](https://cloud.ibm.com/docs/hs-crypto?topic=hs-crypto-get-started) for this resource.",
 									},
+									isInstanceTemplateVolAttTags: {
+										Type:        schema.TypeSet,
+										Optional:    true,
+										ForceNew:    true,
```
Comment (hkantare) [inline ibm/service/vpc/resource_ibm_is_instance_template.go:248]: why forcenew to true we don't have update support?

File: ibm/service/vpc/resource_ibm_is_instance_template.go [hunk 2]
```diff
@@ -437,6 +448,14 @@ func ResourceIBMISInstanceTemplate() *schema.Resource {
 							Optional: true,
 							Computed: true,
 						},
+						isInstanceTemplateBootVolumeTags: {
+							Type:        schema.TypeSet,
```
Comment (hkantare) [inline ibm/service/vpc/resource_ibm_is_instance_template.go:452]: Do we have update support here?

File: ibm/service/vpc/resource_ibm_is_instance_template.go [hunk 3]
```diff
@@ -239,6 +242,14 @@ func ResourceIBMISInstanceTemplate() *schema.Resource {
 										ForceNew:    true,
 										Description: "The CRN of the [Key Protect Root Key](https://cloud.ibm.com/docs/key-protect?topic=key-protect-getting-started-tutorial) or [Hyper Protect Crypto Service Root Key](https://cloud.ibm.com/docs/hs-crypto?topic=hs-crypto-get-started) for this resource.",
 									},
+									isInstanceTemplateVolAttTags: {
+										Type:        schema.TypeSet,
+										Optional:    true,
+										ForceNew:    true,
+										Elem:        &schema.Schema{Type: schema.TypeString, ValidateFunc: validate.InvokeValidator("ibm_is_volume", "tags")},
```
Comment (hkantare) [inline ibm/service/vpc/resource_ibm_is_instance_template.go:249]: validate.InvokeValidator("ibm_is_volume", "tags") should be 
validate.InvokeValidator("ibm_is_instance_template", "tags")...the first argument should be resource name .
And add in the next func if this validator is not added..

Check in all resources

**Comment by hkantare** [comment]
Test existing  basic testcases also  and share the results

File: ibm/service/vpc/data_source_ibm_is_instance_template.go [hunk 4]
```diff
@@ -597,6 +614,9 @@ func dataSourceIBMISInstanceTemplateRead(context context.Context, d *schema.Reso
 					volProfInst := volProfIntf.(*vpcv1.VolumeProfileIdentity)
 					bootVol[isInstanceTemplateBootProfile] = volProfInst.Name
 				}
+				if instance.BootVolumeAttachment.Volume.UserTags != nil {
```
[the remark this answers is not in the document]
Comment (SunithaGudisagarIBM) [reply ibm/service/vpc/data_source_ibm_is_instance_template.go:617]: All the nill checks have happened in the above if lines..
```
		if instance.BootVolumeAttachment != nil {
			bootVolList := make([]map[string]interface{}, 0)
			bootVol := map[string]interface{}{}

			bootVol[isInstanceTemplateDeleteVol] = *instance.BootVolumeAttachment.DeleteVolumeOnInstanceDelete
			if instance.BootVolumeAttachment.Volume != nil {
				volumeIntf := instance.BootVolumeAttachment.Volume
				bootVol[isInstanceTemplateName] = volumeIntf.Name
				bootVol[isInstanceTemplateVol] = volumeIntf.Name
				bootVol[isInstanceTemplateBootSize] = volumeIntf.Capacity
				if instance.BootVolumeAttachment.Volume.Profile != nil {
					volProfIntf := instance.BootVolumeAttachment.Volume.Profile
					volProfInst := volProfIntf.(*vpcv1.VolumeProfileIdentity)
					bootVol[isInstanceTemplateBootProfile] = volProfInst.Name
				}
				if instance.BootVolumeAttachment.Volume.UserTags != nil {
					bootVol[isInstanceTemplateBootVolumeTags] = instance.BootVolumeAttachment.Volume.UserTags
				}
			}
			bootVolList = append(bootVolList, bootVol)
			d.Set(isInstanceTemplateBootVolumeAttachment, bootVolList)
		}
```

File: ibm/service/vpc/resource_ibm_is_instance_template.go (hunk 1 above)
Comment (SunithaGudisagarIBM) [reply ibm/service/vpc/resource_ibm_is_instance_template.go:248]: No harini, other fields in the same block has the same structure. 
https://pages.github.ibm.com/vpc/vpc-spec-artifacts/branch/master/public/2022-03-29.html#/Instances/update_instance_template

```
						isInstanceTemplateVolAttVolPrototype: {
							Type:     schema.TypeList,
							MaxItems: 1,
							MinItems: 1,
							Optional: true,
							ForceNew: true,
							Elem: &schema.Resource{
								Schema: map[string]*schema.Schema{
									isInstanceTemplateVolAttVolIops: {
										Type:        schema.TypeInt,
										Optional:    true,
										ForceNew:    true,
										Description: "The maximum I/O operations per second (IOPS) for the volume.",
									},
.
.
..
.
.

									isInstanceTemplateVolAttTags: {
										Type:        schema.TypeSet,
										Optional:    true,
										ForceNew:    true,
										Elem:        &schema.Schema{Type: schema.TypeString, ValidateFunc: validate.InvokeValidator("ibm_is_volume", "tags")},
										Set:         flex.ResourceIBMVPCHash,
										Description: "UserTags for the volume instance",
									},
								},
							},
						},
```

File: ibm/service/vpc/resource_ibm_is_instance_template.go (hunk 2 above)
Comment (SunithaGudisagarIBM) [reply ibm/service/vpc/resource_ibm_is_instance_template.go:452]: https://pages.github.ibm.com/vpc/vpc-spec-artifacts/branch/master/public/2022-03-29.html#/Instances/update_instance_template
We dont have update support here..

**Comment by SunithaGudisagarIBM** [comment]
make testacc TEST=./ibm/service/vpc TESTARGS='-run=TestAccIBMISInstance_basic'
make testacc TEST=./ibm/service/vpc TESTARGS='-run=TestAccIBMISInstanceTemplate_basic'
make testacc TEST=./ibm/service/vpc TESTARGS='-run=TestAccIBMISInstanceVolumeAttachment_basic'


Screenshot 2022-09-08 at 6 35 33 PM
Screenshot 2022-09-08 at 6 39 09 PM
Screenshot 2022-09-08 at 6 49 48 PM

**Pull request merged**